File: services/events-service/internal/repository/event_repository.go
package repository

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/AnathanWang/andexevents/services/events-service/internal/model"
)

// EventRepository интерфейс репозитория событий
type EventRepository interface {
	Create(ctx context.Context, event *model.Event) error
	GetByID(ctx context.Context, id string) (*model.Event, error)
	GetAll(ctx context.Context, req *model.GetEventsRequest) ([]model.Event, int, error)
	GetByUserID(ctx context.Context, userID string, page, pageSize int) ([]model.Event, int, error)
	Update(ctx context.Context, id string, req *model.UpdateEventRequest) (*model.Event, error)
	Delete(ctx context.Context, id string) error
	GetNearby(ctx context.Context, lat, lon, radius float64, page, pageSize int) ([]model.Event, int, error)
	GetCreatorByEventID(ctx context.Context, eventID string) (*model.EventCreator, error)
}

type eventRepository struct {
	pool *pgxpool.Pool
}

// NewEventRepository создаёт новый репозиторий событий
func NewEventRepository(pool *pgxpool.Pool) EventRepository {
	return &eventRepository{pool: pool}
}

// Create создаёт новое событие
func (r *eventRepository) Create(ctx context.Context, event *model.Event) error {
	query := `
		INSERT INTO "Event" (
			id, title, description, category, location, latitude, longitude,
			"locationGeo", "dateTime", "endDateTime", price, "imageUrl", "isOnline",
			status, "maxParticipants", "minAge", "maxAge", "createdById", "createdAt", "updatedAt"
		) VALUES (
			$1, $2, $3, $4, $5, $6, $7,
			CASE WHEN $6 IS NOT NULL AND $7 IS NOT NULL 
				THEN ST_SetSRID(ST_MakePoint($7, $6), 4326)::geography 
				ELSE NULL 
			END,
			$8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $18
		)
		RETURNING "createdAt", "updatedAt"
	`

	err := r.pool.QueryRow(ctx, query,
		event.ID,
		event.Title,
		event.Description,
		event.Category,
		event.Location,
		event.Latitude,
		event.Longitude,
		event.DateTime,
		event.EndDateTime,
		event.Price,
		event.ImageURL,
		event.IsOnline,
		event.Status,
		event.MaxParticipants,
		event.MinAge,
		event.MaxAge,
		event.CreatedByID,
		time.Now(),
	).Scan(&event.CreatedAt, &event.UpdatedAt)

	return err
}

// GetByID получает событие по ID
func (r *eventRepository) GetByID(ctx context.Context, id string) (*model.Event, error) {
	query := `
		SELECT 
			id, title, description, category, location, latitude, longitude,
			"dateTime", "endDateTime", price, "imageUrl", "isOnline", status,
			"maxParticipants", "minAge", "maxAge", "createdById", "createdAt", "updatedAt"
		FROM "Event"
		WHERE id = $1
	`

	var event model.Event
	err := r.pool.QueryRow(ctx, query, id).Scan(
		&event.ID, &event.Title, &event.Description, &event.Category,
		&event.Location, &event.Latitude, &event.Longitude,
		&event.DateTime, &event.EndDateTime, &event.Price, &event.ImageURL,
		&event.IsOnline, &event.Status, &event.MaxParticipants,
		&event.MinAge, &event.MaxAge, &event.CreatedByID,
		&event.CreatedAt, &event.UpdatedAt,
	)

	if err == pgx.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	return &event, nil
}

// GetAll получает список событий с фильтрацией и пагинацией
func (r *eventRepository) GetAll(ctx context.Context, req *model.GetEventsRequest) ([]model.Event, int, error) {
	baseQuery := `FROM "Event" WHERE 1=1`
	countQuery := `SELECT COUNT(*) ` + baseQuery

	var conditions []string
	var args []interface{}
	argNum := 1

	if req.Category != "" {
		conditions = append(conditions, fmt.Sprintf("category = $%d", argNum))
		args = append(args, req.Category)
		argNum++
	}

	if req.Status != nil {
		conditions = append(conditions, fmt.Sprintf("status = $%d", argNum))
		args = append(args, *req.Status)
		argNum++
	}

	if req.IsOnline != nil {
		conditions = append(conditions, fmt.Sprintf(`"isOnline" = $%d`, argNum))
		args = append(args, *req.IsOnline)
		argNum++
	}

	if req.Search != "" {
		conditions = append(conditions, fmt.Sprintf("(title ILIKE $%d OR description ILIKE $%d)", argNum, argNum))
		args = append(args, "%"+req.Search+"%")
		argNum++
	}

	if len(conditions) > 0 {
		condStr := " AND " + strings.Join(conditions, " AND ")
		baseQuery += condStr
		countQuery += condStr
	}

	var total int
	err := r.pool.QueryRow(ctx, countQuery, args...).Scan(&total)
	if err != nil {
		return nil, 0, err
	}

	orderBy := `"dateTime"`
	switch req.SortBy {
	case "createdAt":
		orderBy = `"createdAt"`
	case "title":
		orderBy = "title"
	}

	orderDir := "ASC"
	if strings.ToLower(req.SortOrder) == "desc" {
		orderDir = "DESC"
	}

	offset := (req.Page - 1) * req.PageSize

	selectQuery := fmt.Sprintf(`
		SELECT 
			id, title, description, category, location, latitude, longitude,
			"dateTime", "endDateTime", price, "imageUrl", "isOnline", status,
			"maxParticipants", "minAge", "maxAge", "createdById", "createdAt", "updatedAt"
		%s
		ORDER BY %s %s
		LIMIT $%d OFFSET $%d
	`, baseQuery, orderBy, orderDir, argNum, argNum+1)

	args = append(args, req.PageSize, offset)

	rows, err := r.pool.Query(ctx, selectQuery, args...)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	var events []model.Event
	for rows.Next() {
		var event model.Event
		err := rows.Scan(
			&event.ID, &event.Title, &event.Description, &event.Category,
			&event.Location, &event.Latitude, &event.Longitude,
			&event.DateTime, &event.EndDateTime, &event.Price, &event.ImageURL,
			&event.IsOnline, &event.Status, &event.MaxParticipants,
			&event.MinAge, &event.MaxAge, &event.CreatedByID,
			&event.CreatedAt, &event.UpdatedAt,
		)
		if err != nil {
			return nil, 0, err
		}
		events = append(events, event)
	}

	return events, total, nil
}

// GetNearby получает события поблизости с использованием PostGIS
func (r *eventRepository) GetNearby(ctx context.Context, lat, lon, radius float64, page, pageSize int) ([]model.Event, int, error) {
	pointQuery := `ST_SetSRID(ST_MakePoint($1, $2), 4326)::geography`

	countQuery := fmt.Sprintf(`
		SELECT COUNT(*)
		FROM "Event"
		WHERE "locationGeo" IS NOT NULL
		AND ST_DWithin("locationGeo", %s, $3)
	`, pointQuery)

	var total int
	err := r.pool.QueryRow(ctx, countQuery, lon, lat, radius).Scan(&total)
	if err != nil {
		return nil, 0, err
	}

	offset := (page - 1) * pageSize
	selectQuery := fmt.Sprintf(`
		SELECT 
			id, title, description, category, location, latitude, longitude,
			"dateTime", "endDateTime", price, "imageUrl", "isOnline", status,
			"maxParticipants", "minAge", "maxAge", "createdById", "createdAt", "updatedAt",
			ST_Distance("locationGeo", %s) as distance
		FROM "Event"
		WHERE "locationGeo" IS NOT NULL
		AND ST_DWithin("locationGeo", %s, $3)
		ORDER BY distance ASC
		LIMIT $4 OFFSET $5
	`, pointQuery, pointQuery)

	rows, err := r.pool.Query(ctx, selectQuery, lon, lat, radius, pageSize, offset)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	var events []model.Event
	for rows.Next() {
		var event model.Event
		var distance float64
		err := rows.Scan(
			&event.ID, &event.Title, &event.Description, &event.Category,
			&event.Location, &event.Latitude, &event.Longitude,
			&event.DateTime, &event.EndDateTime, &event.Price, &event.ImageURL,
			&event.IsOnline, &event.Status, &event.MaxParticipants,
			&event.MinAge, &event.MaxAge, &event.CreatedByID,
			&event.CreatedAt, &event.UpdatedAt, &distance,
		)
		if err != nil {
			return nil, 0, err
		}
		event.Distance = &distance
		events = append(events, event)
	}

	return events, total, nil
}

// GetByUserID получает события созданные пользователем
func (r *eventRepository) GetByUserID(ctx context.Context, userID string, page, pageSize int) ([]model.Event, int, error) {
	countQuery := `SELECT COUNT(*) FROM "Event" WHERE "createdById" = $1`
	var total int
	err := r.pool.QueryRow(ctx, countQuery, userID).Scan(&total)
	if err != nil {
		return nil, 0, err
	}

	offset := (page - 1) * pageSize
	selectQuery := `
		SELECT 
			id, title, description, category, location, latitude, longitude,
			"dateTime", "endDateTime", price, "imageUrl", "isOnline", status,
			"maxParticipants", "minAge", "maxAge", "createdById", "createdAt", "updatedAt"
		FROM "Event"
		WHERE "createdById" = $1
		ORDER BY "createdAt" DESC
		LIMIT $2 OFFSET $3
	`

	rows, err := r.pool.Query(ctx, selectQuery, userID, pageSize, offset)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	var events []model.Event
	for rows.Next() {
		var event model.Event
		err := rows.Scan(
			&event.ID, &event.Title, &event.Description, &event.Category,
			&event.Location, &event.Latitude, &event.Longitude,
			&event.DateTime, &event.EndDateTime, &event.Price, &event.ImageURL,
			&event.IsOnline, &event.Status, &event.MaxParticipants,
			&event.MinAge, &event.MaxAge, &event.CreatedByID,
			&event.CreatedAt, &event.UpdatedAt,
		)
		if err != nil {
			return nil, 0, err
		}
		events = append(events, event)
	}

	return events, total, nil
}

// Update обновляет событие
func (r *eventRepository) Update(ctx context.Context, id string, req *model.UpdateEventRequest) (*model.Event, error) {
	var setClauses []string
	var args []interface{}
	argNum := 1

	if req.Title != nil {
		setClauses = append(setClauses, fmt.Sprintf("title = $%d", argNum))
		args = append(args, *req.Title)
		argNum++
	}
	if req.Description != nil {
		setClauses = append(setClauses, fmt.Sprintf("description = $%d", argNum))
		args = append(args, *req.Description)
		argNum++
	}
	if req.Category != nil {
		setClauses = append(setClauses, fmt.Sprintf("category = $%d", argNum))
		args = append(args, *req.Category)
		argNum++
	}

	if len(setClauses) == 0 {
		return r.GetByID(ctx, id)
	}

	setClauses = append(setClauses, fmt.Sprintf(`"updatedAt" = $%d`, argNum))
	args = append(args, time.Now())
	argNum++

	args = append(args, id)

	query := fmt.Sprintf(`
		UPDATE "Event"
		SET %s
		WHERE id = $%d
		RETURNING id, title, description, category, location, latitude, longitude,
			"dateTime", "endDateTime", price, "imageUrl", "isOnline", status,
			"maxParticipants", "minAge", "maxAge", "createdById", "createdAt", "updatedAt"
	`, strings.Join(setClauses, ", "), argNum)

	var event model.Event
	err := r.pool.QueryRow(ctx, query, args...).Scan(
		&event.ID, &event.Title, &event.Description, &event.Category,
		&event.Location, &event.Latitude, &event.Longitude,
		&event.DateTime, &event.EndDateTime, &event.Price, &event.ImageURL,
		&event.IsOnline, &event.Status, &event.MaxParticipants,
		&event.MinAge, &event.MaxAge, &event.CreatedByID,
		&event.CreatedAt, &event.UpdatedAt,
	)

	if err == pgx.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	return &event, nil
}

// Delete удаляет событие
func (r *eventRepository) Delete(ctx context.Context, id string) error {
	query := `DELETE FROM "Event" WHERE id = $1`
	result, err := r.pool.Exec(ctx, query, id)
	if err != nil {
		return err
	}

	if result.RowsAffected() == 0 {
		return pgx.ErrNoRows
	}

	return nil
}

// GetCreatorByEventID получает информацию о создателе события
func (r *eventRepository) GetCreatorByEventID(ctx context.Context, eventID string) (*model.EventCreator, error) {
	query := `
		SELECT u.id, u.name, u.email, u."photoURL"
		FROM "User" u
		JOIN "Event" e ON e."createdById" = u.id
		WHERE e.id = $1
	`

	var creator model.EventCreator
	err := r.pool.QueryRow(ctx, query, eventID).Scan(
		&creator.ID, &creator.Name, &creator.Email, &creator.PhotoURL,
	)

	if err == pgx.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	return &creator, nil
}

File: services-java/events-service/src/main/java/com/andexevents/events/repo/EventRepository.java
package com.andexevents.events.repo;

import com.andexevents.events.model.EventDtos;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.time.Instant;
import java.util.*;

@Repository
public class EventRepository {
    private final JdbcTemplate jdbcTemplate;

    public EventRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    public String insertEvent(EventCreateParams p) {
        String id = UUID.randomUUID().toString();

        jdbcTemplate.update(
                "INSERT INTO events.\"Event\" (id, title, description, category, location, latitude, longitude, \"locationGeo\", \"dateTime\", \"endDateTime\", price, \"imageUrl\", \"isOnline\", status, \"createdById\", \"createdAt\", \"updatedAt\") " +
                        "VALUES (?, ?, ?, ?, ?, ?, ?, ST_SetSRID(ST_MakePoint(?, ?), 4326)::geography, ?, ?, ?, ?, ?, 'APPROVED'::events.\"EventStatus\", ?, NOW(), NOW())",
                id,
                p.title(),
                p.description(),
                p.category(),
                p.location(),
                p.latitude(),
                p.longitude(),
                p.longitude(),
                p.latitude(),
                Timestamp.from(p.dateTime()),
                p.endDateTime() != null ? Timestamp.from(p.endDateTime()) : null,
                p.price() == null ? 0.0 : p.price(),
                p.imageUrl(),
                p.isOnline() != null && p.isOnline(),
                p.createdById()
        );

        return id;
    }

    public Optional<EventRow> findEventRowById(String id) {
        List<EventRow> rows = jdbcTemplate.query(
                "SELECT * FROM events.\"Event\" WHERE id = ?",
                (rs, rn) -> mapEventRow(rs),
                id
        );
        return rows.stream().findFirst();
    }

    public Optional<EventDtos.CreatorDto> findCreatorByEventId(String eventId) {
        List<EventDtos.CreatorDto> rows = jdbcTemplate.query(
                "SELECT u.id, u.\"displayName\", u.\"photoUrl\" FROM users.\"User\" u JOIN events.\"Event\" e ON e.\"createdById\" = u.id WHERE e.id = ?",
                (rs, rn) -> new EventDtos.CreatorDto(rs.getString("id"), rs.getString("displayName"), rs.getString("photoUrl")),
                eventId
        );
        return rows.stream().findFirst();
    }

    public long countParticipants(String eventId) {
        Long v = jdbcTemplate.queryForObject(
                "SELECT COUNT(1) FROM events.\"Participant\" WHERE \"eventId\" = ?",
                Long.class,
                eventId
        );
        return v == null ? 0 : v;
    }

    public boolean isParticipating(String eventId, String userId) {
        Integer v = jdbcTemplate.query(
                "SELECT 1 FROM events.\"Participant\" WHERE \"eventId\" = ? AND \"userId\" = ? LIMIT 1",
                rs -> rs.next() ? 1 : 0,
                eventId,
                userId
        );
        return v != null && v == 1;
    }

    public String getParticipationStatus(String eventId, String userId) {
        return jdbcTemplate.query(
                "SELECT status FROM events.\"Participant\" WHERE \"eventId\" = ? AND \"userId\" = ? LIMIT 1",
                rs -> rs.next() ? rs.getString("status") : null,
                eventId,
                userId
        );
    }

    public List<EventDtos.ParticipantDto> findTopParticipants(String eventId, int limit) {
        return jdbcTemplate.query(
                "SELECT p.id, p.\"userId\", p.\"eventId\", p.status, p.\"joinedAt\", p.\"updatedAt\", " +
                        "u.id as u_id, u.\"displayName\" as u_displayName, u.\"photoUrl\" as u_photoUrl, u.email as u_email " +
                        "FROM events.\"Participant\" p JOIN users.\"User\" u ON u.id = p.\"userId\" " +
                        "WHERE p.\"eventId\" = ? ORDER BY p.\"joinedAt\" DESC LIMIT ?",
                (rs, rn) -> mapParticipant(rs),
                eventId,
                limit
        );
    }

        public List<EventDtos.ParticipantDto> findParticipants(String eventId) {
        return jdbcTemplate.query(
            "SELECT p.id, p.\"userId\", p.\"eventId\", p.status, p.\"joinedAt\", p.\"updatedAt\", " +
                "u.id as u_id, u.\"displayName\" as u_displayName, u.\"photoUrl\" as u_photoUrl, u.email as u_email " +
                "FROM events.\"Participant\" p JOIN users.\"User\" u ON u.id = p.\"userId\" " +
                "WHERE p.\"eventId\" = ? ORDER BY p.\"joinedAt\" DESC",
            (rs, rn) -> mapParticipant(rs),
            eventId
        );
        }

    public List<EventRow> listApprovedEvents() {
        return jdbcTemplate.query(
                "SELECT * FROM events.\"Event\" WHERE status = 'APPROVED'::events.\"EventStatus\" ORDER BY \"createdAt\" DESC",
                (rs, rn) -> mapEventRow(rs)
        );
    }

    public List<EventRow> listUserApprovedEvents(String userId) {
        return jdbcTemplate.query(
                "SELECT * FROM events.\"Event\" WHERE status = 'APPROVED'::events.\"EventStatus\" AND \"createdById\" = ? ORDER BY \"createdAt\" DESC",
                (rs, rn) -> mapEventRow(rs),
                userId
        );
    }

        public List<EventRow> listUserParticipatedApprovedEvents(String userId) {
        return jdbcTemplate.query(
            "SELECT e.* FROM events.\"Event\" e " +
                "JOIN events.\"Participant\" p ON p.\"eventId\" = e.id " +
                "WHERE p.\"userId\" = ? AND e.status = 'APPROVED'::events.\"EventStatus\" " +
                "ORDER BY p.\"joinedAt\" DESC",
            (rs, rn) -> mapEventRow(rs),
            userId
        );
        }

    public NearbyQueryResult listNearbyApprovedEvents(double lat, double lon, int maxDistanceMeters, String category, int page, int limit) {
        int offset = (page - 1) * limit;

        String categoryClause = (category == null || category.isBlank()) ? "" : " AND e.category = ? ";
        List<Object> params = new ArrayList<>();
        params.add(lon);
        params.add(lat);
        params.add(maxDistanceMeters);
        if (!categoryClause.isEmpty()) {
            params.add(category);
        }
        params.add(limit);
        params.add(offset);

        String sql =
                "SELECT e.*, " +
                        "ST_Distance(e.\"locationGeo\", ST_SetSRID(ST_MakePoint(?, ?), 4326)::geography) as distance, " +
                        "u.id as u_id, u.\"displayName\" as u_displayName, u.\"photoUrl\" as u_photoUrl, " +
                        "COUNT(p.id) as \"participantCount\" " +
                        "FROM events.\"Event\" e " +
                        "LEFT JOIN users.\"User\" u ON e.\"createdById\" = u.id " +
                        "LEFT JOIN events.\"Participant\" p ON e.id = p.\"eventId\" " +
                        "WHERE e.status = 'APPROVED'::events.\"EventStatus\" AND e.\"isOnline\" = false " +
                        "AND ST_DWithin(e.\"locationGeo\", ST_SetSRID(ST_MakePoint(?, ?), 4326)::geography, ?) " +
                        categoryClause +
                        "GROUP BY e.id, u.id " +
                        "ORDER BY distance ASC " +
                        "LIMIT ? OFFSET ?";

        // careful: we used lon/lat twice: for distance and for within
        // params currently contains only once; rebuild to match placeholders:
        List<Object> finalParams = new ArrayList<>();
        finalParams.add(lon);
        finalParams.add(lat);
        finalParams.add(lon);
        finalParams.add(lat);
        finalParams.add(maxDistanceMeters);
        if (!categoryClause.isEmpty()) {
            finalParams.add(category);
        }
        finalParams.add(limit);
        finalParams.add(offset);

        List<NearbyEventRow> events = jdbcTemplate.query(sql, nearbyRowMapper(), finalParams.toArray());

        String countSql =
                "SELECT COUNT(DISTINCT e.id) " +
                        "FROM events.\"Event\" e " +
                        "WHERE e.status = 'APPROVED'::events.\"EventStatus\" AND e.\"isOnline\" = false " +
                        "AND ST_DWithin(e.\"locationGeo\", ST_SetSRID(ST_MakePoint(?, ?), 4326)::geography, ?) " +
                        categoryClause;

        List<Object> countParams = new ArrayList<>();
        countParams.add(lon);
        countParams.add(lat);
        countParams.add(maxDistanceMeters);
        if (!categoryClause.isEmpty()) countParams.add(category);

        Integer total = jdbcTemplate.queryForObject(countSql, Integer.class, countParams.toArray());
        return new NearbyQueryResult(events, total == null ? 0 : total);
    }

    public Optional<EventRow> updateEvent(String eventId, EventUpdateParams p) {
        // For simplicity use update + re-read
        jdbcTemplate.update(
                "UPDATE events.\"Event\" SET title = COALESCE(?, title), description = COALESCE(?, description), category = COALESCE(?, category), " +
                        "location = COALESCE(?, location), latitude = COALESCE(?, latitude), longitude = COALESCE(?, longitude), " +
                        "\"locationGeo\" = CASE WHEN ? IS NOT NULL AND ? IS NOT NULL THEN ST_SetSRID(ST_MakePoint(?, ?), 4326)::geography ELSE \"locationGeo\" END, " +
                        "\"dateTime\" = COALESCE(?, \"dateTime\"), \"endDateTime\" = COALESCE(?, \"endDateTime\"), price = COALESCE(?, price), \"imageUrl\" = COALESCE(?, \"imageUrl\"), \"isOnline\" = COALESCE(?, \"isOnline\"), \"updatedAt\" = NOW() " +
                        "WHERE id = ?",
                p.title(),
                p.description(),
                p.category(),
                p.location(),
                p.latitude(),
                p.longitude(),
                p.latitude(),
                p.longitude(),
                p.longitude(),
                p.latitude(),
                p.dateTime() == null ? null : Timestamp.from(p.dateTime()),
                p.endDateTime() == null ? null : Timestamp.from(p.endDateTime()),
                p.price(),
                p.imageUrl(),
                p.isOnline(),
                eventId
        );
        return findEventRowById(eventId);
    }

    public boolean deleteEvent(String eventId) {
        int affected = jdbcTemplate.update("DELETE FROM events.\"Event\" WHERE id = ?", eventId);
        return affected > 0;
    }

    public Optional<ParticipantRow> upsertParticipation(String eventId, String userId, String status) {
        String id = UUID.randomUUID().toString();
        List<ParticipantRow> rows = jdbcTemplate.query(
                "INSERT INTO events.\"Participant\" (id, \"userId\", \"eventId\", status, \"joinedAt\", \"updatedAt\") " +
                "VALUES (?, ?, ?, ?::events.\"ParticipantStatus\", NOW(), NOW()) " +
                        "ON CONFLICT (\"userId\", \"eventId\") DO UPDATE SET status = EXCLUDED.status, \"updatedAt\" = NOW() " +
                        "RETURNING id, \"userId\", \"eventId\", status, \"joinedAt\", \"updatedAt\"",
                (rs, rn) -> mapParticipantRow(rs),
                id,
                userId,
                eventId,
                status
        );
        return rows.stream().findFirst();
    }

    public Optional<ParticipantRow> deleteParticipation(String eventId, String userId) {
        List<ParticipantRow> rows = jdbcTemplate.query(
                "DELETE FROM events.\"Participant\" WHERE \"eventId\" = ? AND \"userId\" = ? RETURNING id, \"userId\", \"eventId\", status, \"joinedAt\", \"updatedAt\"",
                (rs, rn) -> mapParticipantRow(rs),
                eventId,
                userId
        );
        return rows.stream().findFirst();
    }

    public record EventCreateParams(
            String title,
            String description,
            String category,
            String location,
            double latitude,
            double longitude,
            Instant dateTime,
            Instant endDateTime,
            Double price,
            String imageUrl,
            Boolean isOnline,
            String createdById
    ) {
    }

    public record EventUpdateParams(
            String title,
            String description,
            String category,
            String location,
            Double latitude,
            Double longitude,
            Instant dateTime,
            Instant endDateTime,
            Double price,
            String imageUrl,
            Boolean isOnline
    ) {
    }

    public record EventRow(
            String id,
            String title,
            String description,
            String category,
            String location,
            double latitude,
            double longitude,
            Instant dateTime,
            Instant endDateTime,
            double price,
            String imageUrl,
            boolean isOnline,
            String status,
            String rejectionReason,
            Integer maxParticipants,
            Integer minAge,
            Integer maxAge,
            String createdById,
            Instant createdAt,
            Instant updatedAt
    ) {
    }

    public record NearbyEventRow(EventRow event, double distance, EventDtos.CreatorDto createdBy, long participantCount) {
    }

    public record NearbyQueryResult(List<NearbyEventRow> events, int total) {
    }

    public record ParticipantRow(String id, String userId, String eventId, String status, Instant joinedAt, Instant updatedAt) {
    }

    private EventRow mapEventRow(ResultSet rs) throws SQLException {
        return new EventRow(
                rs.getString("id"),
                rs.getString("title"),
                rs.getString("description"),
                rs.getString("category"),
                rs.getString("location"),
                rs.getDouble("latitude"),
                rs.getDouble("longitude"),
                toInstant(rs.getObject("dateTime")),
                toInstant(rs.getObject("endDateTime")),
                rs.getDouble("price"),
                rs.getString("imageUrl"),
                rs.getBoolean("isOnline"),
                rs.getString("status"),
                rs.getString("rejectionReason"),
                (Integer) rs.getObject("maxParticipants"),
                (Integer) rs.getObject("minAge"),
                (Integer) rs.getObject("maxAge"),
                rs.getString("createdById"),
                toInstant(rs.getObject("createdAt")),
                toInstant(rs.getObject("updatedAt"))
        );
    }

    private RowMapper<NearbyEventRow> nearbyRowMapper() {
        return (rs, rn) -> {
            EventRow event = mapEventRow(rs);
            double distance = rs.getDouble("distance");
            long participantCount = rs.getLong("participantCount");
            EventDtos.CreatorDto createdBy = null;
            String creatorId = rs.getString("u_id");
            if (creatorId != null) {
                createdBy = new EventDtos.CreatorDto(creatorId, rs.getString("u_displayName"), rs.getString("u_photoUrl"));
            }
            return new NearbyEventRow(event, distance, createdBy, participantCount);
        };
    }

    private EventDtos.ParticipantDto mapParticipant(ResultSet rs) throws SQLException {
        EventDtos.UserPreviewDto u = new EventDtos.UserPreviewDto(
                rs.getString("u_id"),
                rs.getString("u_displayName"),
                rs.getString("u_photoUrl"),
                rs.getString("u_email")
        );

        return new EventDtos.ParticipantDto(
                rs.getString("id"),
                rs.getString("userId"),
                rs.getString("eventId"),
                rs.getString("status"),
                toInstant(rs.getObject("joinedAt")),
                toInstant(rs.getObject("updatedAt")),
                u
        );
    }

    private ParticipantRow mapParticipantRow(ResultSet rs) throws SQLException {
        return new ParticipantRow(
                rs.getString("id"),
                rs.getString("userId"),
                rs.getString("eventId"),
                rs.getString("status"),
                toInstant(rs.getObject("joinedAt")),
                toInstant(rs.getObject("updatedAt"))
        );
    }

    private Instant toInstant(Object ts) {
        if (ts == null) return null;
        if (ts instanceof java.sql.Timestamp t) return t.toInstant();
        if (ts instanceof java.time.OffsetDateTime odt) return odt.toInstant();
        if (ts instanceof java.time.LocalDateTime ldt) return ldt.atZone(java.time.ZoneOffset.UTC).toInstant();
        return null;
    }
}
